addons/stock_account/wizard/stock_valuation_layer_revaluation.py

```diff
@@ -0,0 +1,139 @@
+    new_value = fields.Monetary("New value", compute='_compute_new_value')
+    new_value_by_qty = fields.Monetary("New value by quantity", compute='_compute_new_value')
+    reason = fields.Char("Reason", help="Reason of the revaluation")
+
+    account_journal_id = fields.Many2one('account.journal', "Journal", required=True, check_company=True)
+    account_id = fields.Many2one('account.account', "Counterpart Account", domain=[('deprecated', '=', False)], required=True, check_company=True)
+    date = fields.Date("Accounting Date")
+
+    @api.depends('current_value_svl', 'current_quantity_svl', 'added_value')
+    def _compute_new_value(self):
+        for reval in self:
+            reval.new_value = reval.current_value_svl + reval.added_value
+            if not self.company_id.currency_id.is_zero(reval.current_quantity_svl):
+                reval.new_value_by_qty = reval.new_value / reval.current_quantity_svl
+            else:
+                reval.new_value_by_qty = 0.0
+
+    def action_validate_revaluation(self):
+        """ Revaluate the stock for `self.product_id` in `self.company_id`.
+
+        - Change the stardard price with the new valuation by product unit.
+        - Create a manual stock valuation layer with the `added_value` of `self`.
+        - If the Inventory Valuation of the product category is automated, create
+        related account move.
+        """
+        self.ensure_one()
+        if self.currency_id.is_zero(self.added_value):
+            raise UserError(_("The added value doesn't have any impact after rounding"))
+
+        remaining_svls = self.env['stock.valuation.layer'].with_company(self.company_id).search([
+            ('product_id', '=', self.product_id.id),
+            ('remaining_qty', '>', 0),
+            ('company_id', '=', self.company_id.id),
+        ])
+
+        # Create a manual stock valuation layer
+        svl_vals = {
+            'company_id': self.company_id.id,
+            'product_id': self.product_id.id,
+            'description': _("Manual Stock Valuation: %s.%s") % (
+                self.reason or _("No Reason Given"),
+                self.product_id.categ_id.property_cost_method == 'average' and (_(" Product cost move from %s to %s") % (self.product_id.standard_price, self.new_value_by_qty)) or ''),
+            'value': self.added_value,
+            'quantity': 0,
+        }
+
+        remaining_quantity = sum(remaining_svls.mapped('remaining_qty'))
+        added_value_remaining = self.added_value
+        for slv in remaining_svls:
+            remaining_value_to_add = self.currency_id.round(self.added_value * slv.remaining_qty / remaining_quantity)
+            slv.remaining_value += remaining_value_to_add
+            added_value_remaining -= remaining_value_to_add
+        if not float_is_zero(added_value_remaining, precision_rounding=self.currency_id.rounding):
+            remaining_svls[-1].remaining_value += added_value_remaining
```

Comment:
i don't like the idea of looping on remianing_svl and then accessing the last element out of the loop
can you try something similar to?

``` python
        remaining_qty = sum(remaining_svls.mapped('remaining_qty'))
        remaining_value = self.added_value
        remaining_value_unit_cost = self.currency_id.round(remaining_value / remaining_qty)
        for svl in remaining_svls:
            if float_is_zero(svl.remaining_qty - remaining_qty):
                svl.remaining_value += remaining_value
            else:
                taken_remaining_value = unit_cost * svl.remaining_qty
                svl.remaining_value += taken_remaining_value
                remaining_value -= taken_remaining_value
                remaining_qty -= svl.remaining_qty
```

and still see if the 0.33 0.33 0.34 usecase works? thanks